core: using native interrupt handlers (#192)
OP-TEE/optee_docs: architecture/core.rst

several trusted threads in optee_os helps normal world scheduler to be
efficient.

Core handlers for native interrupts
===================================

OP-TEE core offers means to implement native interrupt
(see :ref:`native_foreign_irqs`) controller drivers and means for device
drivers to configure and register handler functions for these interrupts.
Interrupt handlers can be nested for when an interrupt controller
exposes interrupts which signaling is multiplexed on an interrupt
controlled by a parent interrupt controller.

Interrupt controllers are represented by an instance of ``struct itr_chip``.
An interrupt controller exposes a given number of interrupts, each identified
by an index from 0 to N-1 where N is the total number of interrupts exposed

> by that controller. In the literature, an interrupt index identifier

Reviewer: litterature

is called interrupt number.

The root interrupt controller currently implemented in OP-TEE core is a
GIC instance for the Arm architecture.

**Interrupt management API functions**

Interrupt management resources are declared in header file

> kernel/interrupt.h_. Main API function are:

Reviewer: ``` ``<kernel/interrupt.h>`` ``` maybe?

Reviewer: Yes fine with me.

    - ``interrupt_enable()`` and ``interrupt_disable()`` to respectively

>       enable and disable and interrupt.

Reviewer: an

    - ``interrupt_mask()`` and ``interrupt_unmask()`` to respectively

>       mask/unmask and interrupt.

Reviewer: an

    - ``interrupt_configure()`` to configure an interrupt detection mode
      and the interrupt priority.

    - ``interrupt_add_handler()`` and ``interrupt_remove_handler()`` to
      respectively register and unregister an interrupt handler function
      for a given interrupt.

When configuration switch ``CFG_DRIVERS_INTERRUPT`` is enabled, core
offers a means to use Device Tree to describe interrupts topologyin the
platform. When used, the platform must provide a secure devicetree blob

> (DTB) which platform and device drivers can use to getreferences to

Reviewer: get references

> interrupt controllers and relative interrupt numbers theyconsume.

Reviewer: they consume

> Main API functions are:

Reviewer: The main (some definite articles are missing in a few other places I believe but my English is not good enough to decide it omitting them is OK or not).

    - ``dt_register_interrupt_provider()`` to register an interrupt

>       controller related to a given node in the Device Tree. This function

Reviewer: Shouldn't the function be called `dt_register_interrupt_controller()` then?

Author: I understand your point of view. However all functions used the register a provider of device driver reference in _dt_drvier.c_ (and friends) resources are labelled `dt_register_xxx_provider()` so I would prefer to keep that `_provider` suffix. Here I could propose `dt_register_interrupt_controller_provider()` but it's a bit long, almost as long  as existing `dt_driver_device_from_node_idx_prop_phandle()` which I feel pretty bad about :(

What about `dt_register_itr_chip_provider()`?

Reviewer: Ah, consistency again ;) it's fine to keep `dt_register_interrupt_provider()` then.

      requires, among its arguments, a callback function that is called when
      an interrupt consumer requires a reference to a specific interrupt
      exposed by the controller.

    - ``dt_get_interrupt()``, ``dt_get_interrupt_by_index()`` and
      ``dt_get_interrupt_by_name()`` for an interrupt consumer to get the
      reference to the interrupt controller and number it consumes, based
      on the properties found in the consumer Device Tree node.

**Interrupt controller drivers**

Interrupt controller instance (``struct itr_chip``) registers operation
function handlers for management of the interrupt. An interrupt controller
must provide operation handler functions ``add``, ``enable`` and ``disable``.
There are other operation handler functions but these are not mandatory, as

> ``mask``, ``unmask``, ``rasie_pi``, ``raise_sgi`` and ``set_priority``.

Reviewer: raise
Open question: should parentheses be used when documenting function pointers?

Operation handler function ``add`` is called when an interrupt consumer
registers a function handler for a given interrupt.

A reference to CPU root interrupt controller, e.g. a GIC instance on Arm
architecture CPUs, can be retrieved with API function ``itr_core_get()``.

**Interrrupt handlers**

Interrupt handler functions are functions registered by drivers that

> core shall call when the related interrupt occurs. Structure struct itr_handler

Reviewer: ``` ``struct itr_handler`` ```

references a handler. It contains the handler function entry point as well
as interrupt information: its number, its interrupt controller device and

> a few fields.

Reviewer: a few other fields

An interrupt handler function return value is of type ``enum itr_return``.
It shall return ``ITRR_HANDLED`` when the interrupt is served and
``ITRR_NONE`` when the interrupt cannot be served.

The interrupt handler executes in an interrupt context, not in a thread

> context. The handler must therefore execute very quickly. If there are

Reviewer: "as quickly as possible"?
Maybe say why it is so (what happens when a handler runs for a long time)

    };

    static enum itr_return foo_it_handler_fn(struct itr_handler *h)
    {
            foo_acknowledge_interrupt(h->it);.
            DMSG("Interrupt FOO%u served", h->it);

            return ITRR_HANDLED;
    }

    static TEE_Result foo_initialization(void)
    {
            TEE_Result res = TEE_ERROR_GENERIC;

            res = interrupt_alloc_add_handler(itr_core_get(),
                                              GIC_INT_FOO,
                                              foo_it_handler_fn,
                                              ITRF_TRIGGER_LEVEL,
                                              &foo_int1_data,
                                              &foo_int1_handler);
            if (res)
                    return res;

            interrupt_enable(itr_chip, it_num);

            return TEE_SUCCESS;
    }

    static void foo_release(void)
    {
            if (foo_int1_handler) {
                    interrupt_disale(foo_int1_handler->chip,
                                     foo_int1_handler->it);

                    interrupt_remove_free_handler(&foo_int1_handler);
            }
    }

**Using device tree**

> Device drivers can rely on Device Tree technology to get references to the

Reviewer: I would remove "technology"

Author: ok

several trusted threads in optee_os helps normal world scheduler to be
efficient.

Core handlers for native interrupts
===================================

> OP-TEE core offers means to implement native interrupt
> (see :ref:`native_foreign_irqs`) controller drivers and means for device
> drivers to configure and register handler functions for these interrupts.

Reviewer 2: ```suggestion
OP-TEE core provides methods for device drivers to setup and register
handler functions for native interrupt controller drivers
(see:ref:`native_foreign_irqs`).
```

Interrupt handlers can be nested for when an interrupt controller
exposes interrupts which signaling is multiplexed on an interrupt
controlled by a parent interrupt controller.

Interrupt controllers are represented by an instance of ``struct itr_chip``.
An interrupt controller exposes a given number of interrupts, each identified
by an index from 0 to N-1 where N is the total number of interrupts exposed

> by that controller. In the litterature, an interrupt index identifier

Reviewer 2: s/litterature/literature/

Author: @jforissier and I should take my English lessons :)

Reviewer 2: To be fair, I'm just as much native English speaker as you are 😄

Author: But you're still a good reference for me. :beers:

is called interrupt number.

The root interrupt controller currently implemented in OP-TEE core is a
GIC instance for the Arm architecture.

**Interrupt management API functions**

Interrupt management resources are declared in header file
kernel/interrupt.h_. Main API function are:

    - ``interrupt_enable()`` and ``interrupt_disable()`` to respectively
      enable and disable an interrupt.

    - ``interrupt_mask()`` and ``interrupt_unmask()`` to respectively
      mask/unmask an interrupt.

    - ``interrupt_configure()`` to configure an interrupt detection mode
      and priority.

    - ``interrupt_add_handler()`` and ``interrupt_remove_handler()`` to
      respectively register and unregister an interrupt handler function
      for a given interrupt.

> When configuration switch ``CFG_DRIVERS_INTERRUPT`` is enabled, core
> offers a means to use Device Tree to describe interrupts topologyin the
> platform. When used, the platform must provide a secure devicetree blob

Reviewer 2: There is a spelling error "topologyin", but I suggest rewriting the sentence slightly.
```suggestion
When the configuration switch ``CFG_DRIVERS_INTERRUPT`` is enabled,
the core provides a way to describe the platform's interrupt topology
using Device Tree.
```

Reviewer 2: The "When used, ..." should be there. It's just the one sentence I had in mind.

Author: Thanks. I will also replace `CFG_DRIVERS_INTERRUPT` with `CFG_DT`. `CFG_DRIVERS_INTERRUPT` is useless.

(DTB) which platform and device drivers can use to get references to
interrupt controllers and relative interrupt numbers they consume.
The Main API functions are:

    - ``dt_register_interrupt_provider()`` to register an interrupt
      controller related to a given node in the Device Tree. This function
      requires, among its arguments, a callback function that is called when
      an interrupt consumer requires a reference to a specific interrupt
      exposed by the controller.

    - ``dt_get_interrupt()``, ``dt_get_interrupt_by_index()`` and
      ``dt_get_interrupt_by_name()`` for an interrupt consumer to get the
      reference to the interrupt controller and number it consumes, based
      on the properties found in the consumer Device Tree node.

**Interrupt controller drivers**

> An interrupt controller instance (``struct itr_chip``) registers operation
> function handlers for management of the interrupt it controls. An interrupt
> controller
> must provide operation handler functions ``.add``, ``.enable`` and ``.disable``.
> There are other operation handler functions but these are not mandatory, as
> ``.mask``, ``.unmask``, ``.rasie_pi``, ``.raise_sgi`` and ``.set_priority``.

Reviewer 2: Text not changed in the suggestion below, it's just that the word "controller" was alone on an entire line.

```suggestion
An interrupt controller instance (``struct itr_chip``) registers operation
function handlers for management of the interrupt it controls. An interrupt
controller must provide operation handler functions ``.add``, ``.enable`` and
``.disable``. There are other operation handler functions but these are not
mandatory, as ``.mask``, ``.unmask``, ``.rasie_pi``, ``.raise_sgi`` and
``.set_priority``.
```

A reference to CPU root interrupt controller, e.g. a GIC instance on Arm
architecture CPUs, can be retrieved with API function ``itr_core_get()``.

**Interrrupt handlers**

Interrupt handler functions are functions registered by drivers that
core shall call when the related interrupt occurs. Structure
``struct itr_handler``

> references a handler. It contains the handler function entry point as well
> as interrupt information: its number, its interrupt controller device and
> a few other fields.

Reviewer 2: ```suggestion
references a handler. It contains the handler function entry point as well
as interrupt information such as the interrupt number, interrupt controller
device, and a few more parameters.
```

Author: thanks. I've slightly rephrased the information.

An interrupt handler function return value is of type ``enum itr_return``.
It shall return ``ITRR_HANDLED`` when the interrupt is served and
``ITRR_NONE`` when the interrupt cannot be served.

> The interrupt handler executes in an interrupt context, not in a thread
> context. This means all other interrupts are masked in the mean time,
> the handler must therefore execute as fast as possible. If there are

Reviewer 2: Suggestion for the sentence starting with "The interrupt" to the next sentence that ends with "... fast as possible".

```suggestion
The interrupt handler runs in an interrupt context rather than a thread
context. When this occurs, all other interrupts are masked, necessitating fast
execution of the interrupt handler to avoid delaying or missing out on other
interrupts.
```

> The interrupt handler executes in an interrupt context, not in a thread
> context. This means all other interrupts are masked in the mean time,
> the handler must therefore execute as fast as possible. If there are

Reviewer 2: Suggestion for the sentence starting with "The interrupt" to the next sentence that ends with "... fast as possible".

```suggestion
The interrupt handler runs in an interrupt context rather than a thread
context. When this occurs, all other interrupts are masked, necessitating fast
execution of the interrupt handler to avoid delaying or missing out on other
interrupts.
```

For example, the dummy driver below prints a debug trace when the related
interrupt occurs:

.. code-block:: c

    static struct itr_handler *foo_int1_handler;

    static struct foo_int1_data = {
            /* field with some interrupt handler private data */
    };

    static enum itr_return foo_it_handler_fn(struct itr_handler *h)
    {
            foo_acknowledge_interrupt(h->it);.
            DMSG("Interrupt FOO%u served", h->it);

            return ITRR_HANDLED;
    }

    static TEE_Result foo_initialization(void)
    {
            TEE_Result res = TEE_ERROR_GENERIC;

            res = interrupt_alloc_add_handler(itr_core_get(),
                                              GIC_INT_FOO,
                                              foo_it_handler_fn,
                                              ITRF_TRIGGER_LEVEL,
                                              &foo_int1_data,
                                              &foo_int1_handler);
            if (res)
                    return res;

            interrupt_enable(itr_chip, it_num);

            return TEE_SUCCESS;
    }

    static void foo_release(void)
    {
            if (foo_int1_handler) {

>                     interrupt_disale(foo_int1_handler->chip,

Reviewer 3: interrupt_disable

                                     foo_int1_handler->it);

                    interrupt_remove_free_handler(&foo_int1_handler);
            }
    }

**Using device tree**

Device drivers can rely on Device Tree to get references to the

> interrupt they consume. Refer to Device Tree document

Reviewer 2: "document" or "documentation"?

Author: indeed
